### DISCUSSION TITLE
Support for nested data

### DISCUSSION THREAD

**Discussion by scosco** [apache/superset#25328] [Ideas]
**Problems**

Arrays of structs/objects are displayed as arrays of arrays which is very confusing because some DBs allow arrays within arrays. Also the subcolumn names are hidden unless table is flattened and new fields are queried.

Download only as flat csv, no json/parquet/etc that would support nested and/or repeated data (i.e. arrays of structs).

**Solution**

- Display arrays of objects correctly with their sub-column (keys) names in the result/table preview and in the schema.
- Add download option for json/parquet/... in case result table is nested.
- Add json/nesting support for data visualization

**Example UI: Bigquery**
In a DBS without arrays/nesting this would be two tables: basket and fruits.
Ideally Superset would be able to surface data and schema like this:
![image](https://github.com/apache/superset/assets/12495829/9028cb10-c566-40fe-bbfd-841cb2f7639f)
![image](https://github.com/apache/superset/assets/12495829/ca7a0703-b329-4b70-a67a-cb460a8309c5)
![image](https://github.com/apache/superset/assets/12495829/9bca31d8-263b-4964-8099-419842b6915d)

**Why/Impact/Value**

DBs that have functionality for arrays and nesting allow to effectively pre-join data, keeping together events that happened with their repeated attributes (e.g. a purchase event containing all products bought and their configuration). This reduces analysis mistakes and avoids overhead data like PK/FK. Postgres, Trino, BigQuery, Spark all have an UNNEST or similar operator turning arrays into relations and when containing objects/structs then turning those into columns - so the idea of repeated/nested data is widely implemented. 

Being able to utilize this paradigm in Superset would be very helpful since it is also tailored to analytical use cases, not transactional ones.

**Comment by thedeceptio** [comment]
This seems extremely useful. Would make my life a lot easier.

**Comment by markebrown** [comment]
If I am not mistaken, this effectively allows for symmetric aggregates like is available in Looker also, meaning you can sum up at the nested/repeated level as well as the top level and get the correct sum for both for example order and order line item.

**Comment by progressive-scaler** [comment]
I would love to have this features on Superset.

Can we get any feeback of some kind in order to understand if this is is being analyzed to see if it can be done?

Something similar was discussed [here](https://github.com/apache/superset/discussions/18435),

**Comment by thedeceptio** [comment]
I've been running ClickHouse + Superset in production and have run into the same limitation.

I've been exploring a design where array columns become first-class semantic types with an MVP consisting only of:

- Contains
- Explode (arrayJoin/UNNEST)
- Length

The implementation would introduce a generic multi-valued dimension abstraction rather than a ClickHouse-specific feature.

Would maintainers be interested in a SIP/prototype for this?
